src/backend/distributed/executor/executor.c

```diff
@@ -1301,6 +1327,75 @@ ManageWorkerPool(WorkerPool *workerPool)
 }
 
 
+/*
+ * CheckConnectionTimeout makes sure that the execution enforces the connection
+ * establishment timeout defined by the user (NodeConnectionTimeout).
+ *
+ * The rule is that if a worker pool has already ready tasks but no active connections
+ * at all, take an action. For the types of actions, see the comments in the function.
+ *
+ * Enforcing the timeout per pool (over per session) helps the execution to continue
+ * even if we can establish a single connection as we expect to have target pool size
+ * number of connections. In the end, the executor is capable of using one connection
+ * to execute multiple tasks.
+ */
+static void
+CheckConnectionTimeout(WorkerPool *workerPool)
+{
+	DistributedExecution *execution = workerPool->distributedExecution;
+	TimestampTz firstTaskAssignmentTime = workerPool->firstTaskAssignmentTime;
+	TimestampTz now = GetCurrentTimestamp();
+
```

Comment:
might be nice to return early if we already managed to establish a connection before the timeout

Reply:
Hmm, due to `ForceMaxQueryParallelization ` checks, it becomes slightly complicated. But I think compilers could optimize this fairly well, so shall I keep as-is?

Reply:
I added the early exit check now.

Reply:
Oh no, the check I added is different than what you ask. I could maybe move `CheckConnectionTimeout(workerPool);` after the first 3 if checks in `ManageWorkerPool` instead of where it is now. That'd decrease the frequency of calling the function a lot?